text/0000-rustdoc-json.md

```diff
@@ -0,0 +1,1124 @@
+
+### Resolving IDs
+
+The crate's `index` contains mostly local items, which includes impls of external traits on local
+types or local traits on external types. The exception to this is that external trait definitions
+and their associated items are also included in the `index` because this information is useful when
+generating the comprehensive list of methods for a type.
+
+This means that many IDs aren't included in the `index` (any reference to a struct, macro, etc.
+from a different crate). In these cases the fallback is to look up the ID in the crate's `paths`.
+That gives [enough information](#ItemSummary) about the item to create cross references or simply
+provide a name without copying all of the information about external items into the local
+crate's JSON output.
+
+### ExternalCrate
+
+Name      | Type    | Description
+----------|---------|------------------------------------------------------------------------------
+`name`    | String  | The name of the crate.
+`html_root_url` | String  | (*Optional*) The `html_root_url` for that crate if they specify one.
+
+### ItemSummary
+
+Name      | Type    | Description
+----------|---------|------------------------------------------------------------------------------
+`crate_num` | int   | A number corresponding to the crate this Item is from. Used as an key to the `extern_crates` map in [Crate](#Crate). A value of zero represents an Item from the local crate, any other number means that this Item is external.
+`path`    | [String] | The fully qualified path (e.g. `["std", "io", "lazy", "Lazy"]` for `std::io::lazy::Lazy`) of this Item.
+`kind`    | String  | What type of Item this is (see [Item](#Item)).
+
+## Item
+
+An Item represents anything that can hold documentation - modules, structs, enums, functions,
+traits, type aliases, and more. The Item data type holds fields that can apply to any of these,
+and leaves kind-specific details (like function args or enum variants) to the `inner` field.
+
+Name      | Type    | Description
+----------|---------|------------------------------------------------------------------------------
+`crate_num` | int   | A number corresponding to the crate this Item is from. Used as an key to the `extern_crates` map in [Crate](#Crate). A value of zero represents an Item from the local crate, any other number means that this Item is external.
+`name`    | String  | The name of the Item, if present. Some Items, like impl blocks, do not have names.
+`span`    | [Span](#Span) | (*Optional*) The source location of this Item.
+`visibility` | String | `"default"`, `"public"`, `"crate"`, or `"restricted"` (`pub(path)`).
+`restricted_path` | String | (*Optional*) If `visitility == "restricted"`, this field contains the path that it's restricted to.
+`docs`    | String  | The extracted documentation text from the Item.
+`attrs`   | [String] | The attributes (other than doc comments) on the Item, rendered as strings.
```

Comment:
What does 'rendered as strings' mean? Will `#[inline]` be rendered as `"inline"` or `"#[inline]"`?

Reply:
The latter is what it currently gives just because that's what the ast pretty printer outputs. Do you think stripping the `#[]` makes sense? I guess if a consumer is gonna go through all the functions looking for a certain attribute its easier if it's just the string.

Reply:
We should not stabilize anything depending on the AST pretty-printer. That makes me concerned that there are other ambiguities here - what will `#[cfg(feature = "x")]` print out? Will multiple spaces be normalized to one? Will `"` be escaped? What happens if you have both `/// doc comment` and `#[doc = "doc comment"]`?

I'm concerned that we're accidentally locking ourselves into an unstable format that's easy to break. I'd almost rather get rid of `attrs` altogether.

Reply:
The intention of this is to give renderers an "easy path" way to consume this format by giving them a reasonable string of text to present. If we document that the exact format is unstable and subject to change, would that alleviate your concerns @jyn514?